skaffold-plugins-core/src/main/java/com/google/cloud/tools/skaffold/command/Skaffold.java

```diff
@@ -0,0 +1,219 @@
+      CACHED_SKAFFOLD_LOCATION.resolveSibling("skaffold.sha256");
+
+  /**
+   * Initializes {@link Skaffold} with a custom path to the {@code skaffold} executable.
+   *
+   * @param executablePath the path to {@code skaffold}
+   * @return a new {@link Skaffold}
+   */
+  public static Skaffold atPath(Path executablePath) {
+    return new Skaffold(executablePath.toString());
+  }
+
+  /**
+   * Initializes {@link Skaffold} with a managed {@code skaffold} executable.
+   *
+   * @return a new {@link Skaffold}
+   * @throws IOException if an I/O exception occurred
+   */
+  public static Skaffold init() throws IOException {
+    ensureSkaffoldIsLatestVersion(CACHED_SKAFFOLD_LOCATION, CACHED_SKAFFOLD_DIGEST_LOCATION);
+    return new Skaffold(CACHED_SKAFFOLD_LOCATION.toString());
+  }
+
+  @VisibleForTesting
+  static void ensureSkaffoldIsLatestVersion(
+      Path cachedSkaffoldLocation, Path cachedSkaffoldDigestLocation) throws IOException {
+    if (Files.exists(cachedSkaffoldLocation)) {
+      // Checks if the digest is up-to-date and redownloads skaffold if not.
+      Path temporaryDigestFile = Files.createTempFile("", "");
+      temporaryDigestFile.toFile().deleteOnExit();
+      SkaffoldDownloader.downloadLatestDigest(temporaryDigestFile);
+      byte[] latestDigest = Files.readAllBytes(temporaryDigestFile);
+      if (Files.exists(cachedSkaffoldDigestLocation)) {
+        byte[] storedDigest = Files.readAllBytes(cachedSkaffoldDigestLocation);
+        if (Arrays.equals(storedDigest, latestDigest)) {
+          return;
+        }
+      }
+      Files.write(cachedSkaffoldDigestLocation, latestDigest);
+    }
+
+    SkaffoldDownloader.downloadLatest(cachedSkaffoldLocation);
+  }
+
+  private static Callable<Void> redirect(InputStream inputStream, OutputStream outputStream) {
+    return () -> {
+      try (InputStream inputStream1 = inputStream) {
+        ByteStreams.copy(inputStream1, outputStream);
+      }
+      return null;
+    };
+  }
+
+  private final List<String> initialTokens;
+
+  private Function<List<String>, ProcessBuilder> processBuilderFactory = ProcessBuilder::new;
+  @Nullable private Path skaffoldYaml;
+  @Nullable private String profile;
+  @Nullable private InputStream stdinInputStream;
+  @Nullable private OutputStream stdoutOutputStream;
+  @Nullable private OutputStream stderrOutputStream;
+
+  @VisibleForTesting
+  Skaffold(String... initialTokens) {
+    this.initialTokens = Arrays.asList(initialTokens);
+  }
+
+  public Skaffold setSkaffoldYaml(Path skaffoldYaml) {
+    this.skaffoldYaml = skaffoldYaml;
+    return this;
+  }
+
+  public Skaffold setProfile(String profile) {
+    this.profile = profile;
+    return this;
+  }
+
+  /**
+   * Sets the {@link InputStream} to provide to {@code skaffold} as the stdin.
+   *
+   * @param stdinInputStream provides the stdin
+   * @return this
+   */
+  public Skaffold setStdinInputStream(InputStream stdinInputStream) {
```

Comment:
Weird naming for `setStdinInputStream()`, `setStdoutOutputStream()`, and `setStderrOutputStream()`.  Consider either ProcessBuilder-style `setInput()`/`setOutput()`, or Process-style `setInputStream()`/`setOutputStream()`?

Reply:
Hmm, these methods are to set the `InputStream`s/`OutputStream`s that provide the stdin/receive the stdout. There are also two setters for `OutputStream`s that receive the stdout and stderr separately.